### PR TITLE
Add DATEFROMPARTS function to SqlServer

### PR DESCRIPTION
**Pull request by Marusyk**
Contributes #19237

Add `DATEFROMPARTS` function

Please review,
Thank you in advance

### CODE REVIEW

File: src/EFCore.SqlServer/Query/Internal/SqlServerDateTime2FromPartsFunctionTranslator.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+using System.Linq;
+using System.Reflection;
+using JetBrains.Annotations;
+using Microsoft.EntityFrameworkCore.Query;
+using Microsoft.EntityFrameworkCore.Query.SqlExpressions;
+using Microsoft.EntityFrameworkCore.Storage;
+
+namespace Microsoft.EntityFrameworkCore.SqlServer.Query.Internal
+{
+    public class SqlServerDateTime2FromPartsFunctionTranslator : IMethodCallTranslator
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/Internal/SqlServerDateTime2FromPartsFunctionTranslator.cs]: We should combine all these method call translators which works on from parts into single class.

File: src/EFCore.SqlServer/Query/Internal/SqlServerDateFromPartsFunctionTranslator.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        }
+
+        public virtual SqlExpression Translate(SqlExpression instance, MethodInfo method, IReadOnlyList<SqlExpression> arguments)
+        {
+            return _methodInfo.Equals(method)
+                ? _sqlExpressionFactory.Function(
+                    "DATEFROMPARTS",
+                    arguments.Skip(1),
+                    _methodInfo.ReturnType,
+                    _typeMappingSource.FindMapping(typeof(DateTime), "datetime"))
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/Internal/SqlServerDateFromPartsFunctionTranslator.cs]: return server type is date.

File: src/EFCore.SqlServer/Query/Internal/SqlServerDateTime2FromPartsFunctionTranslator.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        public virtual SqlExpression Translate(SqlExpression instance, MethodInfo method, IReadOnlyList<SqlExpression> arguments)
+        {
+            return _methodInfo.Equals(method)
+                ? _sqlExpressionFactory.Function(
+                    "DATETIME2FROMPARTS",
+                    arguments.Skip(1),
+                    _methodInfo.ReturnType,
+                    _typeMappingSource.FindMapping(typeof(DateTime), "datetime"))
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/Internal/SqlServerDateTime2FromPartsFunctionTranslator.cs]: return server type is datetime2.

File: src/EFCore.SqlServer/Query/Internal/SqlServerDateTimeOffsetFromPartsFunctionTranslator.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        public virtual SqlExpression Translate(SqlExpression instance, MethodInfo method, IReadOnlyList<SqlExpression> arguments)
+        {
+            return _methodInfo.Equals(method)
+                ? _sqlExpressionFactory.Function(
+                    "DATETIMEOFFSETFROMPARTS",
+                    arguments.Skip(1),
+                    _methodInfo.ReturnType,
+                    _typeMappingSource.FindMapping(typeof(DateTime), "datetime"))
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/Internal/SqlServerDateTimeOffsetFromPartsFunctionTranslator.cs]: return server type is datetimeoffset.

File: src/EFCore.SqlServer/Query/Internal/SqlServerSmallDateTimeFromPartsFunctionTranslator.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        public virtual SqlExpression Translate(SqlExpression instance, MethodInfo method, IReadOnlyList<SqlExpression> arguments)
+        {
+            return _methodInfo.Equals(method)
+                ? _sqlExpressionFactory.Function(
+                    "SMALLDATETIMEFROMPARTS",
+                    arguments.Skip(1),
+                    _methodInfo.ReturnType,
+                    _typeMappingSource.FindMapping(typeof(DateTime), "datetime"))
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/Internal/SqlServerSmallDateTimeFromPartsFunctionTranslator.cs]: smalldatetime

File: src/EFCore.SqlServer/Query/Internal/SqlServerTimeFromPartsFunctionTranslator.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        public virtual SqlExpression Translate(SqlExpression instance, MethodInfo method, IReadOnlyList<SqlExpression> arguments)
+        {
+            return _methodInfo.Equals(method)
+                ? _sqlExpressionFactory.Function(
+                    "TIMEFROMPARTS",
+                    arguments.Skip(1),
+                    _methodInfo.ReturnType,
+                    _typeMappingSource.FindMapping(typeof(DateTime), "datetime"))
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/Internal/SqlServerTimeFromPartsFunctionTranslator.cs]: time

**Comment by maumar** [comment]
@Marusyk please fix the failing test, ideally convert the added tests to AssertQuery pattern and add tests for each method (datetime2, datetimeoffset, smalldatetime, time). Looks good otherwise

**Comment by Marusyk** [comment]
Could anyone suggest?
I have problem with `TimeFromParts` :

```csharp
var count = context.Orders
     .Count(c => new TimeSpan(23, 59, 0) > EF.Functions.TimeFromParts(12, 12, 31, 5, 1));
```
it throws 

> System.InvalidOperationException
>   HResult=0x80131509
>   Message=An exception was thrown while attempting to evaluate the LINQ query parameter expression '(new TimeSpan(23, 59, 0) > EF.Functions.TimeFromParts(12, 12, 31, 5, 1))'.
>   Source=Microsoft.EntityFrameworkCore
>   StackTrace:
>    at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.GetValue(Expression expression, String& parameterName) in C:\Users\rmarusyk\source\EntityFrameworkCore\src\EFCore\Query\Internal\ParameterExtractingExpressionVisitor.cs:line 458
>    at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.Evaluate(Expression expression, Boolean generateParameter) in C:\Users\rmarusyk\source\EntityFrameworkCore\src\EFCore\Query\Internal\ParameterExtractingExpressionVisitor.cs:line 292
>    at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.Visit(Expression expression) in C:\Users\rmarusyk\source\EntityFrameworkCore\src\EFCore\Query\Internal\ParameterExtractingExpressionVisitor.cs:line 110
>    at System.Linq.Expressions.ExpressionVisitor.VisitLambda[T](Expression`1 node)
>    at System.Linq.Expressions.Expression`1.Accept(ExpressionVisitor visitor)
>    at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
>    at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.Visit(Expression expression) in C:\Users\rmarusyk\source\EntityFrameworkCore\src\EFCore\Query\Internal\ParameterExtractingExpressionVisitor.cs:line 113
>    at System.Linq.Expressions.ExpressionVisitor.VisitUnary(UnaryExpression node)
>    at System.Linq.Expressions.UnaryExpression.Accept(ExpressionVisitor visitor)
>    at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
>    at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.Visit(Expression expression) in C:\Users\rmarusyk\source\EntityFrameworkCore\src\EFCore\Query\Internal\ParameterExtractingExpressionVisitor.cs:line 113
>    at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
>    at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
>    at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
>    at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
>    at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.Visit(Expression expression) in C:\Users\rmarusyk\source\EntityFrameworkCore\src\EFCore\Query\Internal\ParameterExtractingExpressionVisitor.cs:line 113
>    at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.ExtractParameters(Expression expression) in C:\Users\rmarusyk\source\EntityFrameworkCore\src\EFCore\Query\Internal\ParameterExtractingExpressionVisitor.cs:line 84
>    at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExtractParameters(Expression query, IParameterValues parameterValues, IDiagnosticsLogger`1 logger, Boolean parameterize, Boolean generateContextAccessors) in C:\Users\rmarusyk\source\EntityFrameworkCore\src\EFCore\Query\Internal\QueryCompiler.cs:line 191
>    at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query) in C:\Users\rmarusyk\source\EntityFrameworkCore\src\EFCore\Query\Internal\QueryCompiler.cs:line 90
>    at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression) in C:\Users\rmarusyk\source\EntityFrameworkCore\src\EFCore\Query\Internal\EntityQueryProvider.cs:line 79
>    at System.Linq.Queryable.Count[TSource](IQueryable`1 source, Expression`1 predicate)
>    at Microsoft.EntityFrameworkCore.Query.NorthwindDbFunctionsQuerySqlServerTest.TimeFromParts_constant_compare() in C:\Users\rmarusyk\source\EntityFrameworkCore\test\EFCore.SqlServer.FunctionalTests\Query\NorthwindDbFunctionsQuerySqlServerTest.cs:line 1005
> 
>   This exception was originally thrown at this call stack:
>     Microsoft.EntityFrameworkCore.SqlServerDbFunctionsExtensions.TimeFromParts(Microsoft.EntityFrameworkCore.DbFunctions, int, int, int, int, int) in SqlServerDbFunctionsExtensions.cs
>     Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.GetValue(System.Linq.Expressions.Expression, out string) in ParameterExtractingExpressionVisitor.cs
> 
> Inner Exception 1:
> InvalidOperationException: The 'TimeFromParts' method is not supported because the query has switched to client-evaluation. Inspect the log to determine which query expressions are triggering client-evaluation.
>

**Comment by smitpatel** [comment]
>  EF.Functions.TimeFromParts(12, 12, 31, 5, 1)

Use at least one argument which is range variable from server side. (i.e. using c)

**Comment by maumar** [comment]
how about:

```cs
var result = context.Orders.Select(o => new { o.OrderID, EF.Functions.TimeFromParts(12, 12, 31, 5, o.OrderId % 60) }).ToList()
```

**Comment by maumar** [comment]
@Marusyk thank you for the contribution!

File: src/EFCore.SqlServer/Query/Internal/SqlServerFromPartsFunctionTranslator.cs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+            [NotNull] IRelationalTypeMappingSource typeMappingSource)
+        {
+            _sqlExpressionFactory = sqlExpressionFactory;
+            _typeMappingSource = typeMappingSource;
+        }
+
+        public virtual SqlExpression Translate(SqlExpression instance, MethodInfo method, IReadOnlyList<SqlExpression> arguments)
+        {
+            if (_dateFromPartsMethodInfo.Equals(method))
+            {
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/Internal/SqlServerFromPartsFunctionTranslator.cs]: refactor to set function name and return server type using case block.

Comment (Marusyk) [reply src/EFCore.SqlServer/Query/Internal/SqlServerFromPartsFunctionTranslator.cs]: Agree. Should I create a new PR for this?

Comment (smitpatel) [reply src/EFCore.SqlServer/Query/Internal/SqlServerFromPartsFunctionTranslator.cs]: I will do it. I may have small clean up tasks in other unrelated PRs too.

**Pull request merged**